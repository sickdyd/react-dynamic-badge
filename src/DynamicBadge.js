
/**
 * This component will trim an array according to the available space in its
 * parent element and if needed add a badge representing the number of hidden items
 * 
 * @param {string} [props.badgeClass] If needed, the user can give a custom class to be used to represent the badges, if not provided use default class
 * @param {Array} props.items Contains an array with all the items to be shown
 * @param {int} [props.minWidth] Sets the minimum width for the text to be shown; smaller than that show the badge
 * @param {bool} [props.onlyBadge] If set to true always show only the badge and no text
 * @param {int} [props.resizeDebounce] Its the value that represents the ms used to debounce the resize event, default 1ms
 * 
 */

import React from 'react';

import './styles/default.css';

import ResizeObserver from 'resize-observer-polyfill';

export default function DynamicBadge(props) {

  // This ref is used to access the width and the fontSize of the container
  const refContainer = React.useRef(null);

  // Contains the jsx state to be rendered
  const [state, setState] = React.useState({
    badgeCounter: 0,
    itemsToDisplay: "",
  });

  // Used to trigger rerenders on window resize
  const [containerWidth, setContainerWidth] = React.useState();

  /**
   * @function getElementContentWidth It returns an int containing the element width
   * @param {DOM Element} element The element used to get the width
   */

  function getElementContentWidth(element) {

    let styles = window.getComputedStyle(element);
    let padding = parseFloat(styles.paddingLeft) + parseFloat(styles.paddingRight);

    return element.clientWidth - padding;
  }

  /**
   * @function getElementFontSize Returns a string containing the element font size
   * @param {DOM Element} element The element used to retrieve the fontSize
   */

  function getElementFontSize(element) {

    let styles = window.getComputedStyle(element);

    return styles.fontSize;
  }

  React.useEffect(()=>{

    const resizeDebounce = props.resizeDebounce ? props.resizeDebounce : 1;

    // Used to debounce the resize event
    function debounce(fn, ms) {

      let timer
      return _ => {
        clearTimeout(timer)
        timer = setTimeout(_ => {
          timer = null
          fn.apply(this, arguments)
        }, ms)
      };
    }

    // Called everytime the the parent element changes size
    // Put a delay between updates of X
    const debouncedHandleResize = debounce( function handleResize() {

      let rulerSpan = document.querySelector(".bdg-ruler-span");

      if (rulerSpan) rulerSpan.style.fontSize = getElementFontSize( refContainer.current.parentNode )

      if ( refContainer.current ) {
        setContainerWidth( getElementContentWidth( refContainer.current.parentNode ));
      }
    }, resizeDebounce)

    const resizeObserver = new ResizeObserver( ()=> debouncedHandleResize() );

    resizeObserver.observe( refContainer.current.parentNode );

    return ()=>{
      resizeObserver.disconnect();
    }

  }, [ props.resizeDebounce ]);

  React.useLayoutEffect(()=>{

    let rulerSpan = document.querySelector(".bdg-ruler-span");
    let rulerBadge = document.querySelector(".bdg-ruler-badge");

    /**
     * If they don't exist, create the rulers used to calculate the pixel length of the items and badge
     */

    if (rulerSpan === undefined || rulerBadge === undefined || rulerSpan === null || rulerBadge === null) {

      rulerSpan = document.createElement("span");
      rulerSpan.classList.add("bdg-ruler-span");
      rulerSpan.classList.add("bdg-phantom");
      rulerSpan.style.fontSize = getElementFontSize( refContainer.current.parentNode );
      document.body.appendChild( rulerSpan );
  
      rulerBadge = document.createElement("div");
      rulerBadge.classList.add("bdg-ruler-badge");
      rulerBadge.classList.add("bdg-phantom");
      // It is possible to pass a custom class for the badge
      // by default use the basic badge
      if (props.badgeClass) {
        rulerBadge.classList.add( props.badgeClass );
      } else {
        rulerBadge.classList.add("bdg-badge");
      }
      
      document.body.appendChild( rulerBadge );
    }

    // Store the container width of the parent element containing the items and badge
    const containerW = getElementContentWidth( refContainer.current.parentNode );

    // If the props items is present create 2 new arrays
    // One will be used to iterate and the other to pop items that can't fit
    const items = props.items ?  Array.from( props.items ) : [];
    let popItems = props.items ?  Array.from( props.items ) : [];

    // This represents the width at which only a badge is visible and no text 
    // for example when the column is too small
    let defaultMinWidth = 0;

    // If the array is not empty
    if (items.length > 0) {

      // Set the ruler with text; if there's only one item, just add that and three dots, other wise add also the comma and one character less
      rulerSpan.innerHTML = items.length === 1 ? items[0].substring(0, 3) + "..." : items[0].substring(0, 2) + ", ...";
      // Set the default minimum width that will be used if the user doesn't specify a minimum width
      defaultMinWidth = rulerSpan.offsetWidth;
    }

    // Set the minWidth with props if present
    const minWidth = props.minWidth ? props.minWidth : defaultMinWidth;

    let badgeCounter = 0;

    // If there is only one item, ignore the badge width for calculations
    let badgeWidth = items.length === 1 ? 0 : rulerBadge.offsetWidth;

    // If the container is smaller than the minWidth or props.onlyBadge
    if ((( containerW - badgeWidth ) < ( minWidth )) || props.onlyBadge) {

      // Hide all text
      popItems = [];
      // Just show a badge with the number
      badgeCounter = items.length;
    }

    else
    
    // If the items is only one, no need for badges
    if ( items.length > 0 ) {

      // Set again the badge width for calculations
      badgeWidth = rulerBadge.offsetWidth;

      items.forEach( (item, index)=> {

        // Set a bool to check if it's the last item
        let isLastItem = ( popItems[ popItems.length - 1 ]  === items[ items.length -1 ] );

        // Join all items in a comma separated string
        let textToCheck = popItems.join(", ");

        if ( isLastItem ) {
          // If it's the last item, check if it can fit, but also
          // remove the space required for the badge, that won't be
          // shown since it's the last item
          rulerSpan.innerHTML = textToCheck;
          badgeWidth = 0;
        } else {
          // If it's not the last item, consider badge width
          // and add , .... to the string to check
          // (the extra dot is to give a bit of room before making the item disappear)
          rulerSpan.innerHTML = textToCheck + ", .....";
          badgeWidth = rulerBadge.offsetWidth;
        }

        // Here, for testing purposes the props.testRulerSpanWidth is adjusted accordingly if existent, otherwise use the defaul turlerSpan.offestWidth
        const rulerSpanWidth = props.testRulerSpanWidth ? props.testRulerSpanWidth - (props.testRulerSpanWidth / items.length * index) : rulerSpan.offsetWidth;

        // If the text to be shown is larger that the container
        if ( rulerSpanWidth > ( containerW - badgeWidth )) {

          // If it's not the last item
          if ( popItems.length > 1 ) {
            // Remove the last item
            popItems.pop();
            // Increase the number shown in the badge
            badgeCounter++;
          }
        }

      });

    }

    // Join all items that will be visible (if needed the extra dots will be added by the css rule text-overflow: ellipsis in bdg-ellipsis)
    const itemsToDisplay = popItems.join(", ");

    setState({
      badgeCounter: badgeCounter,
      itemsToDisplay: itemsToDisplay,
    });
    
  }, [ containerWidth, props.badgeClass, props.items, props.spanClass ]);

  return (
    <div ref={refContainer}>
      <div style={{ width: "100%" }}>
        {state.badgeCounter > 0 ?
            <div className="bdg-container">
              { !props.onlyBadge && <div className="bdg-ellipsis">{ state.itemsToDisplay === "" ? "" : state.itemsToDisplay + ", ..." }</div> }
              <div className={ props.badgeClass ? props.badgeClass : "bdg-badge" } title={ props.items && props.items.join(", ") }>{ state.badgeCounter }</div>
            </div>
        :
            <div className="bdg-ellipsis" style={{ width: "100%" }}>{ state.itemsToDisplay }</div>}
      </div>
    </div>
  )
}
