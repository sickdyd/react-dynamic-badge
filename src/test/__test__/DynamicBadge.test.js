import React from 'react'
import renderer from 'react-test-renderer'
import shallow from 'react-test-renderer/shallow'

import DynamicBadge from './../../DynamicBadge.js'

// Testing with the following items

const items = []

for (let i = 0; i < 10; i++) {
  items.push('Item ' + i)
}

beforeEach(() => {
  // Have to mock getComputedStyle() to make it work in tests
  Object.defineProperty(window, 'getComputedStyle', {
    value: () => ({
      paddingRight: 0,
      paddingLeft: 0,
      fontSize: 10
    })
  })
})

describe('<DynamicBadge>', () => {
  test('match shallow snapshot', () => {
    const renderer = new shallow()

    renderer.render(<DynamicBadge items={items} testRulerSpanWidth={100} />)

    const result = renderer.getRenderOutput()

    expect(result).toMatchSnapshot()
  })

  test('match snapshot', () => {
    // Define a mock for the parent element
    const parentNode = { clientWidth: 400 }

    const component = renderer.create(
      <DynamicBadge items={items} testRulerSpanWidth={100} />,

      {
        // This mock is needed for the refContainer
        createNodeMock: (element) => {
          if (element.type === 'div') {
            // mock parent element
            return {
              // Pass the previously created mock parentNode
              parentNode: parentNode
            }
          }
          return null
        }
      }
    )

    expect(component.toJSON()).toMatchSnapshot()
  })

  for (let i = 0; i < items.length + 1; i++) {
    const message =
      i === items.length
        ? 'renders ' + i + ' items and no badge'
        : 'renders ' + i + ' items and a badge with value ' + (items.length - i)

    let component = null

    test(message, () => {
      // Define a mock for the parent element, make it small (1px)
      const parentNode = {
        clientWidth: 100 * i
      }

      const rulerWidth = items.length * 100

      component = renderer.create(
        <DynamicBadge items={items} testRulerSpanWidth={rulerWidth} minWidth={10} />,

        {
          // This mock is needed for the refContainer
          createNodeMock: (element) => {
            if (element.type === 'div') {
              // mock parent element
              return {
                // Pass the previously created mock parentNode
                parentNode: parentNode
              }
            }
            return null
          }
        }
      )

      // Get the instance of the component
      const testInstance = component.root

      if (i < items.length) {
        // Expect the badge to be visible and to contain the number of the items
        expect(testInstance.findByProps({ className: 'bdg-badge' }).children).toEqual([
          (items.length - i).toString()
        ])
      } else {
        // No badge has to be visible
        expect(() => testInstance.findByProps({ className: 'bdg-badge' })).toThrowError(
          /No instances found/
        )
      }
    })
  }

  test('renders only one item', () => {
    // Define a mock for the parent element
    const parentNode = { clientWidth: 400 }

    const component = renderer.create(
      <DynamicBadge items={['Item 1']} />,

      {
        // This mock is needed for the refContainer
        createNodeMock: (element) => {
          if (element.type === 'div') {
            // mock parent element
            return {
              // Pass the previously created mock parentNode
              parentNode: parentNode
            }
          }
          return null
        }
      }
    )

    const testInstance = component.root
    expect(testInstance.findByProps({ className: 'bdg-ellipsis' }).children).toEqual(['Item 1'])
    expect(() => testInstance.findByProps({ className: 'bdg-badge' })).toThrowError(
      /No instances found/
    )
  })

  test('renders only one item in a zero width container, hence showing a badge with value 1', () => {
    // Define a mock for the parent element
    const parentNode = { clientWidth: 0 }

    const component = renderer.create(
      <DynamicBadge items={['Item 1']} testRulerSpanWidth={100} minWidth={10} />,

      {
        // This mock is needed for the refContainer
        createNodeMock: (element) => {
          if (element.type === 'div') {
            // mock parent element
            return {
              // Pass the previously created mock parentNode
              parentNode: parentNode
            }
          }
          return null
        }
      }
    )

    const testInstance = component.root
    expect(testInstance.findByProps({ className: 'bdg-badge' }).children).toEqual(['1'])
  })

  test('renders only the badge', () => {
    // Define a mock for the parent element
    const parentNode = { clientWidth: 400 }

    const component = renderer.create(
      <DynamicBadge items={items} onlyBadge={true} />,

      {
        // This mock is needed for the refContainer
        createNodeMock: (element) => {
          if (element.type === 'div') {
            // mock parent element
            return {
              // Pass the previously created mock parentNode
              parentNode: parentNode
            }
          }
          return null
        }
      }
    )

    const testInstance = component.root
    expect(testInstance.findByProps({ className: 'bdg-badge' }).children).toEqual([
      items.length.toString()
    ])
  })

  test('renders a badge with a custom class', () => {
    // Define a mock for the parent element
    const parentNode = { clientWidth: 400 }

    const component = renderer.create(
      <DynamicBadge items={items} onlyBadge={true} badgeClass="test-badge-class" />,

      {
        // This mock is needed for the refContainer
        createNodeMock: (element) => {
          if (element.type === 'div') {
            // mock parent element
            return {
              // Pass the previously created mock parentNode
              parentNode: parentNode
            }
          }
          return null
        }
      }
    )

    const testInstance = component.root
    expect(testInstance.findByProps({ className: 'test-badge-class' }).children).toEqual([
      items.length.toString()
    ])
  })
})
