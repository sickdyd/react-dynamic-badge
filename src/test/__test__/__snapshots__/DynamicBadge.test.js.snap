// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DynamicBadge> match snapshot 1`] = `
<div>
  <div
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <div
      className="bdg-ellipsis"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      Item 0, Item 1, Item 2, Item 3, Item 4, Item 5, Item 6, Item 7, Item 8, Item 9
    </div>
  </div>
</div>
`;
